### PR TITLE
[libc] delete hdrgen

### LINKED ISSUE #117208: [libc] delete old hdrgen
Now that we have newhdrgen, filing a bug to track the removal of the existing hdrgen (couldn't find an existing bug for this; feel free to dupe this to that if it exists).

I've moved over the buildbots to use newhdrgen (required resetting buildbot's cmakecache).

cc @petrhosek @RoseZhang03 @aaryanshukla

### PR DESCRIPTION
**Pull request by nickdesaulniers**
Thanks to the effort of @RoseZhang03 and @aaryanshukla under the guidance of
@michaelrj-google and @amykhuang, we now have newhdrgen and no longer have a
dependency on TableGen and thus LLVM in order to start bootstrapping a full
build.

This PR removes:
- LIBC_HDRGEN_EXE; the in tree newhdrgen is the only hdrgen that can be used.
- LIBC_USE_NEW_HEADER_GEN; newhdrgen is the default and only option.
- LIBC_HDRGEN_ONLY; there is no need to have a distinct build step for old
  hdrgen.
- libc-api-test and libc-api-test-tidy build targets.
- Deletes all .td files.

It does not rename newhdrgen to just hdrgen. Will follow up with a distinct PR
for that.

Link: #117209
Link: #117254
Fixes: #117208

### CODE REVIEW

File: libc/CMakeLists.txt [hunk 1]
```diff
@@ -54,27 +54,12 @@ set(LIBC_NAMESPACE ${default_namespace}
...
-endif()
 # We will build the GPU utilities if we are not doing a runtimes build.
 option(LIBC_BUILD_GPU_LOADER "Always build the GPU loader utilities" OFF)
 if(LIBC_BUILD_GPU_LOADER OR (LLVM_LIBC_GPU_BUILD AND NOT LLVM_RUNTIMES_BUILD))
   add_subdirectory(utils/gpu)
 endif()
 
-option(LIBC_USE_NEW_HEADER_GEN "Generate header files using new headergen instead of the old one" ON)
-
 set(NEED_LIBC_HDRGEN FALSE)
```
Comment (jhuber6) [inline libc/CMakeLists.txt:63]: This can be deleted?
```suggestion
```

Comment (nickdesaulniers) [reply libc/CMakeLists.txt:63]: Unsure.  Let me check the original context under which it was added.
9dbedcac1243e8e99103bdff37da51dded67b766

File: libc/CMakeLists.txt [hunk 2]
```diff
@@ -88,8 +73,7 @@ if(NOT LLVM_RUNTIMES_BUILD)
     endforeach()
   endif()
 endif()
-option(LIBC_HDRGEN_ONLY "Only build the 'libc-hdrgen' executable" OFF)
-if(LIBC_HDRGEN_ONLY OR NEED_LIBC_HDRGEN)
+if(NEED_LIBC_HDRGEN)
   # When libc is build as part of the runtimes/bootstrap build's CMake run, we
   # only need to build the host tools to build the libc. So, we just do enough
   # to build libc-hdrgen and return.
```
Comment (overmighty) [inline libc/CMakeLists.txt:79]: Shouldn't `NEED_LIBC_HDRGEN` be renamed to something else and the comment below be updated? Might want to add a TODO comment here.

File: libc/CMakeLists.txt (hunk 1 above)
Comment (nickdesaulniers) [reply libc/CMakeLists.txt:63]: also related c4f236e94a4e2df46838a540936bb9b61c771ca1

Comment (jhuber6) [reply libc/CMakeLists.txt:63]: FYI the GPU build still uses that logic to make the `amdhsa-loader` and `nvptx-loader` utilities.

Comment (nickdesaulniers) [reply libc/CMakeLists.txt:63]: right, I think I left the important bits intact in 46251c628e4e4aee38f15419fcc67cc8b6c4d827, PTAL.

File: libc/CMakeLists.txt (hunk 2 above)
Comment (nickdesaulniers) [reply libc/CMakeLists.txt]: 1c11706f1181

File: libc/include/CMakeLists.txt [hunk 3]
```diff
@@ -18,24 +18,14 @@ add_header(
 )
 
 macro(add_header_macro TARGET_NAME YAML_FILE DEF_FILE GEN_HDR DEPENDS)
```
Comment (michaelrj-google) [inline libc/include/CMakeLists.txt:20]: nit: Add a TODO to clean this up

**Review by michaelrj-google — APPROVED** [review]
LGTM from what I can see. I'd like to run the tests locally to double check nothing broke, but I'm very excited to see this land.

File: libc/include/CMakeLists.txt (hunk 3 above)
Comment (nickdesaulniers) [reply libc/include/CMakeLists.txt:22]: ae75b6b1c9b90bdc1335cfdd96b38accc47f248e

I tried replacing invocations of `add_header_macro` with just `add_gen_header`, but this seemed to run into issues. Not sure if every function invocation of `add_gen_header` would need explicit `DEF_FILE` and `GEN_HDR` spelled out?

**Comment by nickdesaulniers** [comment]
The post submit buildbots with the libc tag are all moved to newhdrgen now.  I'll wait for explicit acks from @petrhosek and @jhuber6 before attempting to crash-land this.

File: libc/include/CMakeLists.txt (hunk 3 above)
Comment (RoseZhang03) [reply libc/include/CMakeLists.txt:22]: I believe they will need explicit DEF_FILE and GEN_HDR. The macro for `add_gen_header2` (lines22-29) should be used, but can be renamed to `add_gen_header`

**Review by petrhosek — APPROVED** [review]
We no longer depend on old hdrgen.

**Comment by nickdesaulniers** [comment]
"Test documentation build" looks new to me in presubmit.

> cp: cannot stat 'libc-build/docs/*': No such file or directory

Pretty sure `-DLIBC_INCLUDE_DOCS=ON` needs to be set. The cmake invocation was:

> cmake -B libc-build -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_RUNTIMES="libc" -DLLVM_ENABLE_SPHINX=ON ./runtimes

ugh, I forget where the presubmit testing stuff is located...

**Comment by nickdesaulniers** [comment]
> ugh, I forget where the presubmit testing stuff is located...

https://github.com/llvm/llvm-project/blob/main/.github/workflows/docs.yml#L159-L165

**Review by jhuber6 — COMMENTED** [review]
The GPU loader utilities still need this weird "forced projects" thing.
```diff
diff --git a/libc/CMakeLists.txt b/libc/CMakeLists.txt
index 73ab38227768..fd82359022cf 100644
--- a/libc/CMakeLists.txt
+++ b/libc/CMakeLists.txt
@@ -50,16 +50,28 @@ set(LIBC_NAMESPACE ${default_namespace}
   CACHE STRING "The namespace to use to enclose internal implementations. Must start with '__llvm_libc'."
 )
 
-
-add_subdirectory(newhdrgen)
-
-
 # We will build the GPU utilities if we are not doing a runtimes build.
 option(LIBC_BUILD_GPU_LOADER "Always build the GPU loader utilities" OFF)
-if(LIBC_BUILD_GPU_LOADER OR (LLVM_LIBC_GPU_BUILD AND NOT LLVM_RUNTIMES_BUILD))
-  add_subdirectory(utils/gpu)
+if(LIBC_BUILD_GPU_LOADER OR NOT LLVM_RUNTIMES_BUILD)
+  foreach(_name ${LLVM_RUNTIME_TARGETS})
+    if("libc" IN_LIST RUNTIMES_${_name}_LLVM_ENABLE_RUNTIMES)
+      if("${_name}" STREQUAL "amdgcn-amd-amdhsa" OR "${_name}" STREQUAL "nvptx64-nvidia-cuda")
+        set(LIBC_NEED_LOADER_UTILS TRUE)
+      endif()
+    endif()
+  endforeach()
+  if("${LIBC_TARGET_TRIPLE}" STREQUAL "amdgcn-amd-amdhsa" OR
+     "${LIBC_TARGET_TRIPLE}" STREQUAL "nvptx64-nvidia-cuda")
+    set(LIBC_NEED_LOADER_UTILS TRUE)
+  endif()
+  if(LIBC_NEED_LOADER_UTILS)
+    add_subdirectory(utils/gpu)
+    return()
+  endif()
 endif()
 
+add_subdirectory(newhdrgen)
+
 option(LIBC_CMAKE_VERBOSE_LOGGING
   "Log details warnings and notifications during CMake configuration." OFF)
 
diff --git a/llvm/CMakeLists.txt b/llvm/CMakeLists.txt
index 03c1486c45ee..cb7a8bf6cbf8 100644
--- a/llvm/CMakeLists.txt
+++ b/llvm/CMakeLists.txt
@@ -190,6 +190,10 @@ if("${LIBC_TARGET_TRIPLE}" STREQUAL "amdgcn-amd-amdhsa" OR
    "${LIBC_TARGET_TRIPLE}" STREQUAL "nvptx64-nvidia-cuda")
   set(LLVM_LIBC_GPU_BUILD ON)
 endif()
+if (NOT "libc" IN_LIST LLVM_ENABLE_PROJECTS AND LLVM_LIBC_GPU_BUILD)
+  message(STATUS "Enabling libc project to build libc testing tools")
+  list(APPEND LLVM_ENABLE_PROJECTS "libc")
+endif()
 
 # LLVM_ENABLE_PROJECTS_USED is `ON` if the user has ever used the
 # `LLVM_ENABLE_PROJECTS` CMake cache variable.  This exists for
```
This seemed to work for me.

**Comment by nickdesaulniers** [comment]
> The GPU loader utilities still need this weird "forced projects" thing.

Wait, that should be orthogonal to old vs new hdrgen.  If anything, we probably need to rethink #117265.

**Comment by jhuber6** [comment]
> > The GPU loader utilities still need this weird "forced projects" thing.
> 
> Wait, that should be orthogonal to old vs new hdrgen. If anything, we probably need to rethink #117265.

It's fine, it doesn't build anything `libc` related, just a handful of utilities. The only `libc` thing they depend on is the `printf` converter stuff, if we moved that to a shared location we could just make it an LLVM utility or something.

File: libc/docs/gpu/building.rst [hunk 4]
```diff
@@ -86,16 +86,13 @@ compiler. These tools must all be up-to-date with the libc source.
      -DCMAKE_C_COMPILER=$HOST_C_COMPILER      \
```
Comment (nickdesaulniers) [inline libc/docs/gpu/building.rst:86]: @jhuber6 re-reading https://libc.llvm.org/gpu/building.html, I think we can/should remove `libc` from the `-DLLVM_ENABLE_PROJECTS` list on L85 now that building libc-hdrgen is no longer necessary. WDYT?

Comment (jhuber6) [reply libc/docs/gpu/building.rst:86]: Definitely, I was actually going to bring that up but forgot.

**Comment by nickdesaulniers** [comment]
@jhuber6 I guess with #117265, I'd expect you to start seeing warnings for GPU builds.
https://libc.llvm.org/gpu/building.html specifically mentions `libc` in `DLLVM_ENABLE_PROJECTS`, which as part of the goal of #78479 is to totally avoid any documentation or support for `libc` in `DLLVM_ENABLE_PROJECTS` (only `DLLVM_ENABLE_RUNTIMES` will be supported).

For the diff you suggested, I don't understand the hunk to llvm/CMakeLists.txt; for a runtimes build, I thought llvm/CMakeLists.txt was not invoked? Only libc/CMakeLists.txt? Perhaps I'm mistaken.

If anything, I suspect I should a patch first to relax #117265 for GPU builds, and/or update the GPU docs to avoid any mention of `DLLVM_ENABLE_PROJECTS`+libc.

**Comment by jhuber6** [comment]
I think I looked at it and thought that the warning printed *before* the additional one was added.

**Comment by nickdesaulniers** [comment]
@jhuber6 9807f27c95e2 PTAL

**Comment by nickdesaulniers** [comment]
Maybe we can just delete the section https://libc.llvm.org/gpu/building.html#standalone-cross-build and just support https://libc.llvm.org/gpu/building.html#runtimes-cross-build ? Then there are no more references to libc in DLLVM_ENABLE_PROJECTS (which is the goal of #78479)?

> This method is generally not recommended as it can only target a single GPU architecture.

In which case, I wonder if we can just drop support for that?

**Comment by nickdesaulniers** [comment]
rebasing on top of https://github.com/llvm/llvm-project/commit/ce416a073be6da06be7fd2d427b896b380fb3877 for the presubmit bots...

/me holds breath
/me dies

File: libc/docs/gpu/building.rst (hunk 4 above)
Comment (nickdesaulniers) [reply libc/docs/gpu/building.rst:82]: 4a474332c8f3

**Review by jhuber6 — APPROVED** [review]

**Comment by nickdesaulniers** [comment]
Ack, need to remove `libc-api-test` from the bots. Will do that now.

**Comment by ilovepi** [comment]
Hi, I think something in the Windows configs may have been missed. we're seeing the following CMake error on our Windows bot:
```
CMake Error at C:/b/s/w/ir/x/w/llvm-llvm-project/libc/CMakeLists.txt:359 (message):
  C:/b/s/w/ir/x/w/llvm-llvm-project/libc/config/windows/headers.txt file not
  found and fullbuild requested.


-- Configuring incomplete, errors occurred!
```

Bot link:  https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-windows-x64/b8729538683736875265/overview

Logs: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8729538683736875265/+/u/clang/configure/stdout

I assume there's just some missing CMake line for windows that was missed, but I'm not super familiar w/ the libc build. Do you have an idea of where that would be?

**Comment by ilovepi** [comment]
oh, it looks like you added the headers.txt file for other platforms in this patch, so I guess that just needs to be handled for Windows too.

**Pull request merged**